File: javascript/point.mjs
class Point {
	constructor(x, y) {
		this.x = x;
		this.y = y;
	}
	up() {
		return new Point(this.x, this.y+1);
	}
	right() {
		return new Point(this.x+1, this.y);
	}
	down() {
		return new Point(this.x, this.y-1);
	}
	left() {
		return new Point(this.x-1, this.y);
	}
	toString() {
		return JSON.stringify({
			x: this.x,
			y: this.y,
		});
	}
}

export {
	Point,
};

File: dart/lib/point.dart
class Point {
	final int x;
	final int y;

	Point(this.x, this.y);

	Point up() {
		return Point(x, y + 1);
	}

	Point right() {
		return Point(x + 1, y);
	}

	Point down() {
		return Point(x, y - 1);
	}

	Point left() {
		return Point(x - 1, y);
	}

	@override
	String toString() {
		return '($x, $y)';
	}

	@override
	bool operator ==(other) {
		return (other is Point) && other.x == x && other.y == y;
	}

	@override
	int get hashCode => Object.hash(x, y);
}
